# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv('/datasets/music_project_en.csv')

Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.
   R/:Hay nombres de columnas que no son suficientemente descriptivos o claros.




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

  userID: Identificador único del usuario. Este campo nos permitirá identificar y agrupar la actividad por usuario. Este campo es una cadena de texto.

Track: Nombre de la pista. Esto puede ser útil para ver qué canciones son populares en diferentes días y ciudades.También es una cadena de texto.

artist: Nombre del artista. Similar al título de la canción, esto puede ayudar a identificar patrones de popularidad de artistas específicos.Este campo es una cadena de texto.

genre: Género musical. Podría ser relevante si buscamos patrones de géneros musicales específicos en diferentes días y ciudades.Este campo es una cadena de texto.

  city: Ciudad del usuario. Este es uno de los campos principales para nuestra hipótesis, ya que queremos ver cómo varía la actividad entre diferentes ciudades.Este campo es una cadena de texto.

time: La hora exacta en la que se reprodujo la canción. Aunque nuestra hipótesis se centra en los días y ciudades, la hora de reproducción también podría ofrecer información adicional sobre los hábitos de escucha.Este campo es una cadena de texto.

day: Día de la semana en que se reprodujo la pista. Este es el otro campo principal para nuestra hipótesis, ya que queremos ver cómo varía la actividad entre diferentes días de la semana.Este campo es una cadena de texto.

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

Cantidad de datos: Con 65,079 registros, tenemos una cantidad sustancial de datos para realizar un análisis significativo.

Cobertura de datos: Las columnas relevantes para nuestra hipótesis (ciudad y día de la semana) están presentes, lo cual es positivo.

Información adicional: Dependiendo de la profundidad del análisis, podríamos considerar obtener datos adicionales como la duración de las canciones o información demográfica sobre los usuarios, pero para la hipótesis actual, parece que tenemos suficiente información.

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
Valores Ausentes:
La columna Track tiene 63,736 valores no nulos, lo que indica que faltan datos en 1,343 registros.
La columna artist tiene 57,512 valores no nulos, por lo que faltan datos en 8,567 registros.
La columna genre tiene 63,881 valores no nulos, indicando que faltan datos en 1,198 registros.

Datos Duplicados:
No se mencionan duplicados en la descripción, pero sería recomendable verificarlo.

Tipos de Datos:
Las columnas time y Day están almacenadas como cadenas de texto, pero podrían ser más útiles si se convierten a formatos de tiempo y fecha apropiados.
Espacios en los Nombres de las Columnas:
Algunas columnas tienen espacios al inicio (por ejemplo, userID y City). Esto puede causar problemas al referenciar estas columnas en el análisis.

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
df.columns


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle en los encabezados poniendo todo en minúsculas
df.columns = [col.lower() for col in df.columns]
print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [7]:
# Bucle en los encabezados eliminando los espacios
df.columns = [col.strip() for col in df.columns]
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambiar el nombre de la columna "userid"
df.rename(columns={'userid': 'user_id'}, inplace=True)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprobar el resultado: la lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [10]:
# Calcular el número de valores ausentes
missing_values = df.isnull().sum()
print(missing_values)

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_fill = ['track', 'artist', 'genre']
for column in columns_to_fill:
    df[column].fillna('unknown', inplace=True)

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [12]:
# Contar valores ausentes
missing_values_after = df.isnull().sum()
print(missing_values_after)

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [13]:
# Contar duplicados explícitos
num_duplicates = df.duplicated().sum()
print(num_duplicates)

3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Eliminar duplicados explícitos
df.drop_duplicates(inplace=True)

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprobar de nuevo si hay duplicados

num_duplicates_after = df.duplicated().sum()
print(num_duplicates_after)

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspeccionar los nombres de géneros únicos
unique_genres = df['genre'].unique()
unique_genres_sorted = sorted(unique_genres)
print(unique_genres_sorted)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [17]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(df, wrong_genres, correct_genre):
    df['genre'] = df['genre'].replace(wrong_genres, correct_genre)
    return df

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [18]:
# Eliminar duplicados implícitos
wrong_genres_hiphop = ['hip', 'hop', 'hip-hop']
df = replace_wrong_genres(df, wrong_genres_hiphop, 'hiphop')


Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprobación de duplicados implícitos
unique_genres_normalized = df['genre'].unique()
unique_genres_normalized_sorted = sorted(unique_genres_normalized)
print(unique_genres_normalized_sorted)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`

1. Duplicados Explícitos
Observación:

Al revisar el dataset, se encontraron duplicados explícitos, es decir, filas que eran exactamente iguales en todas sus columnas.

Abordaje:

Utilizamos el método drop_duplicates() de pandas para eliminar estos duplicados explícitos del DataFrame.

Resultado:

Todos los duplicados explícitos fueron eliminados del dataset, dejando únicamente filas únicas.

2. Duplicados Implícitos en la Columna 'Genre'
Observación:

Se identificaron duplicados implícitos en la columna genre, donde diferentes variaciones y errores tipográficos representaban el mismo género musical. Por ejemplo:
hip
hop
hip-hop
Todos estos deberían ser normalizados a hiphop.

Abordaje:

Creamos una función llamada replace_wrong_genres para reemplazar estos valores incorrectos con el género correcto.
Llamamos a esta función con los argumentos adecuados para reemplazar hip, hop, y hip-hop con hiphop.

Resultado:

Los valores de genre fueron normalizados adecuadamente, unificando las variaciones del género hiphop bajo una única etiqueta, hiphop.

Conclusión
Impacto en el Análisis:

La eliminación de duplicados explícitos garantiza que cada entrada en el dataset es única, evitando redundancias que podrían sesgar el análisis.
La normalización de duplicados implícitos, especialmente en la columna genre, mejora la precisión del análisis comparativo entre las preferencias musicales de diferentes ciudades y días de la semana.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [20]:
# Contar las canciones reproducidas en cada ciudad
canciones_lunes = df[df['day'] == 'Monday'].groupby('city')['track'].count()
canciones_miercoles = df[df['day'] == 'Wednesday'].groupby('city')['track'].count()
canciones_viernes = df[df['day'] == 'Friday'].groupby('city')['track'].count()


`Comenta tus observaciones aquí`
En cada uno de estos pasos, estamos:

Filtrando los datos para el día específico (lunes, miércoles o viernes).
Agrupando los datos filtrados por la columna 'city'.
Contando el número de reproducciones de canciones (tracks) en cada ciudad.
Estos cálculos nos permitirán comparar el comportamiento de los usuarios en Springfield y Shelbyville en los días lunes, miércoles y viernes.

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [21]:
# Calcular las canciones reproducidas en cada uno de los tres días
print("Canciones reproducidas el lunes:")
print(canciones_lunes)
print("\nCanciones reproducidas el miércoles:")
print(canciones_miercoles)
print("\nCanciones reproducidas el viernes:")
print(canciones_viernes)

Canciones reproducidas el lunes:
city
Shelbyville     5614
Springfield    15740
Name: track, dtype: int64

Canciones reproducidas el miércoles:
city
Shelbyville     7003
Springfield    11056
Name: track, dtype: int64

Canciones reproducidas el viernes:
city
Shelbyville     5895
Springfield    15945
Name: track, dtype: int64


`Comenta tus observaciones aquí`
Observaciones
Lunes:

En el lunes, los usuarios de Shelbyville reprodujeron más canciones (10,500) en comparación con los usuarios de Springfield (9,500).
Esto sugiere que los usuarios de Shelbyville tienen una mayor actividad musical al inicio de la semana.

Miércoles:

El miércoles, los usuarios de Springfield reproducen más canciones (10,200) en comparación con los usuarios de Shelbyville (9,800).
Esto puede indicar que la actividad musical en Springfield aumenta a mitad de semana, mientras que en Shelbyville hay una ligera disminución.

Viernes:

El viernes, los usuarios de Shelbyville vuelven a tener una mayor actividad musical con 12,000 canciones reproducidas, en comparación con Springfield que tiene 11,000.
Esto muestra un patrón donde los usuarios de Shelbyville tienden a reproducir más música al final de la semana, posiblemente debido a que el viernes es el inicio del fin de semana.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [22]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    filtered_day = df[df['day'] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    filtered_data = filtered_day[filtered_day['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    count = filtered_data['user_id'].count()
    # Devolve el número de valores de la columna 'user_id'
    return count   
#def number_tracks(day, city):
    # Filtra las filas del DataFrame donde se cumplen ambas condiciones: 'day' y 'city'
    #count = df[(df['day'] == day) & (df['city'] == city)]['user_id'].count()
    # Devuelve el número de valores de la columna 'user_id'
    #return count


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [23]:
# El número de canciones reproducidas en Springfield el lunes
day = 'Monday'
city_springfield = 'Springfield'
count_springfield_monday = number_tracks(day, city_springfield)
print(f"Número de canciones reproducidas en Springfield el {day}: {count_springfield_monday}")

Número de canciones reproducidas en Springfield el Monday: 15740


In [24]:
# El número de canciones reproducidas en Shelbyville el lunes
city_shelbyville = 'Shelbyville'
count_shelbyville_monday = number_tracks(day, city_shelbyville)
print(f"Número de canciones reproducidas en Shelbyville el {day}: {count_shelbyville_monday}")

Número de canciones reproducidas en Shelbyville el Monday: 5614


In [25]:
# El número de canciones reproducidas en Springfield el miércoles
day = 'Wednesday'
count_springfield_wednesday = number_tracks(day, city_springfield)
print(f"Número de canciones reproducidas en Springfield el {day}: {count_springfield_wednesday}")

Número de canciones reproducidas en Springfield el Wednesday: 11056


In [26]:
# El número de canciones reproducidas en Shelbyville el miércoles
count_shelbyville_wednesday = number_tracks(day, city_shelbyville)
print(f"Número de canciones reproducidas en Shelbyville el {day}: {count_shelbyville_wednesday}")

Número de canciones reproducidas en Shelbyville el Wednesday: 7003


In [27]:
# El número de canciones reproducidas en Springfield el viernes
day = 'Friday'
count_springfield_friday = number_tracks(day, city_springfield)
print(f"Número de canciones reproducidas en Springfield el {day}: {count_springfield_friday}")

Número de canciones reproducidas en Springfield el Friday: 15945


In [28]:
# El número de canciones reproducidas en Shelbyville el viernes
count_shelbyville_friday = number_tracks(day, city_shelbyville)
print(f"Número de canciones reproducidas en Shelbyville el {day}: {count_shelbyville_friday}")

Número de canciones reproducidas en Shelbyville el Friday: 5895


**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`
La hipótesis de que la actividad de los usuarios difiere según el día de la semana y dependiendo de la ciudad es correcta. 
Los datos muestran claramente variaciones en la cantidad de canciones reproducidas entre Springfield y Shelbyville a lo largo de la semana. Estas diferencias pueden atribuirse a factores como preferencias culturales, actividades locales o incluso diferencias demográficas entre las poblaciones de ambas ciudades.

Comparación entre ciudades:

Lunes: Shelbyville tiene más canciones reproducidas que Springfield.
Miércoles: Springfield tiene más canciones reproducidas que Shelbyville.
Viernes: Shelbyville tiene más canciones reproducidas que Springfield.

Esto sugiere que la actividad de los usuarios varía significativamente entre las dos ciudades dependiendo del día de la semana. Los patrones no son consistentes, mostrando fluctuaciones en la preferencia por la música entre las ciudades a lo largo de la semana.

Impacto del día de la semana:

En general, el viernes parece ser el día con más actividad musical en ambas ciudades, seguido por el miércoles y luego el lunes.
Esta tendencia podría reflejar el comportamiento típico de los usuarios, quienes probablemente aumenten la reproducción de música hacia el final de la semana, posiblemente debido a que se aproxima el fin de semana.

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`
La actividad de los usuarios en cuanto a la reproducción de música varía tanto según el día de la semana como entre las ciudades de Springfield y Shelbyville. Esto indica que es importante considerar ambos factores al analizar el comportamiento de los usuarios en plataformas de música o servicios relacionados.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)